pkg/metrics/metrics_test.go

```diff
@@ -31,6 +34,130 @@ func Test_ObserveUpdateDurationByType(t *testing.T) {
 	m.ObserveUpdateDurationByType("foo", 1)
 }
 
+func Test_ObserveOngoingStateDuration_Success(t *testing.T) {
+	for _, test := range []struct {
+		name          string
+		state         api.State
+		stateDuration time.Duration
+		setStartedAt  bool
+		expectedState api.State
+	}{
+		{
+			name:          "success_with_state_preparing",
+			state:         api.StatePreparing,
+			stateDuration: time.Hour * 2,
+			setStartedAt:  true,
+			expectedState: api.StatePreparing,
+		},
+		{
+			name:          "success_with_state_cleaning",
+			state:         api.StateCleaning,
+			stateDuration: time.Minute * 25,
+			setStartedAt:  true,
+			expectedState: api.StateCleaning,
+		},
+		{
+			name:          "success_when_state_undefined",
+			state:         api.StateUndefined,
+			stateDuration: time.Hour * 2,
+			setStartedAt:  false,
+			expectedState: api.StateNew,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			// SETUP
+			m := NewMetrics().(*metrics)
+			reg := prometheus.NewPedanticRegistry()
+			reg.MustRegister(m.OngoingStateDuration)
+
+			run := fakePipelineRun(test.state, test.stateDuration, test.setStartedAt)
+
+			// EXERCISE
+			resultErr := m.ObserveOngoingStateDuration(run)
+
+			// VERIFY
+			assert.NilError(t, resultErr)
+
+			metricFamily, err := reg.Gather()
+			assert.NilError(t, err)
+			assert.Equal(t, len(metricFamily), 1)
+			assert.Equal(t, len(metricFamily[0].GetMetric()), 1)
+
+			ioMetric := metricFamily[0].GetMetric()[0]
+			assert.Equal(t, ioMetric.Label[0].GetName(), "state")
+			assert.Equal(t, ioMetric.Label[0].GetValue(), string(test.expectedState))
+
+			for _, bucket := range ioMetric.Histogram.Bucket {
+				duration := float64(test.stateDuration.Seconds())
+				if duration <= *bucket.UpperBound {
+					assert.Equal(t, *bucket.CumulativeCount, uint64(1))
+				} else {
+					assert.Equal(t, *bucket.CumulativeCount, uint64(0))
+				}
+			}
+		})
+	}
+}
+
+func Test_ObserveOngoingStateDuration_Failures(t *testing.T) {
+	for _, test := range []struct {
+		name          string
+		state         api.State
+		stateDuration time.Duration
+		setStartedAt  bool
+		expectedError error
```

Comment:
As test parameter it's not required to pass an error. Use string instead:
```suggestion
		expectedError string
```

Of course, adapt other places accordingly.